accounts-db/src/accounts_db.rs

```diff
@@ -2741,85 +2754,95 @@ impl AccountsDb {
     /// 1. one of the pubkeys in the store has account info to a store whose store count is not going to zero
     /// 2. a pubkey we were planning to remove is not removing all stores that contain the account
     fn calc_delete_dependencies(
-        purges: &HashMap<Pubkey, (SlotList<AccountInfo>, RefCount)>,
+        purges: &[RwLock<HashMap<Pubkey, CleaningInfo>>],
         store_counts: &mut HashMap<Slot, (usize, HashSet<Pubkey>)>,
         min_slot: Option<Slot>,
     ) {
         // Another pass to check if there are some filtered accounts which
         // do not match the criteria of deleting all appendvecs which contain them
         // then increment their storage count.
         let mut already_counted = IntSet::default();
-        for (pubkey, (slot_list, ref_count)) in purges.iter() {
-            let mut failed_slot = None;
-            let all_stores_being_deleted = slot_list.len() as RefCount == *ref_count;
-            if all_stores_being_deleted {
-                let mut delete = true;
-                for (slot, _account_info) in slot_list {
-                    if let Some(count) = store_counts.get(slot).map(|s| s.0) {
-                        debug!(
-                            "calc_delete_dependencies()
+        for bin in purges {
+            let bin = bin.read().unwrap();
+            for (
+                pubkey,
+                CleaningInfo {
+                    slot_list,
+                    ref_count,
+                },
+            ) in bin.iter()
```

Comment:
I think we need `in bin.iter().filter(|x| x.1.slot_list.len() != 0)`. Otherwise, we are iterating over all original entries and not just the ones that we previously identified as zero lamports. We might even need `Option` in the CleaningInfo so we know whether to look at this pubkey or not. Previously, the pubkeys that weren't zero lamport were never iterated here. Now ALL pubkeys are iterated here.

Reply:
My bad. Should be fixed now.